File: tests/st/tensor/overload/test_sigmoid.py
# Copyright 2024 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""Test the overload functional method"""
import numpy as np
import pytest
from tests.mark_utils import arg_mark
from tests.st.ops.test_tools.test_op import TEST_OP
from tests.st.utils import test_utils

import mindspore as ms
import mindspore.nn as nn


class SigmoidNet(nn.Cell):
    def construct(self, x):
        return x.sigmoid()


def generate_random_input(shape, dtype):
    return np.random.randn(*shape).astype(dtype)


@test_utils.run_with_cell
def sigmoid_forward_func(x):
    return x.sigmoid()


@arg_mark(plat_marks=['cpu_linux', 'cpu_windows', 'cpu_macos', 'platform_gpu', 'platform_ascend'],
          level_mark='level1',
          card_mark='onecard',
          essential_mark='unessential')
@pytest.mark.parametrize('mode', [ms.GRAPH_MODE, ms.PYNATIVE_MODE])
def test_method_sigmoid(mode):
    """
    Feature: Functional.
    Description: Test functional feature with Tensor.sigmoid.
    Expectation: Run success
    """
    ms.set_context(mode=mode, jit_config={"jit_level": "O0"})
    net = SigmoidNet()
    x = ms.Tensor(np.array([1, 2, 3, 4, 5]), ms.float32)
    output = net(x)
    expect_output = np.array([0.7310586, 0.880797, 0.95257413, 0.98201376, 0.9933072], dtype=np.float32)
    assert np.allclose(output.asnumpy(), expect_output)


@arg_mark(plat_marks=['platform_ascend', 'platform_gpu', 'cpu_linux', 'cpu_windows', 'cpu_macos'],
          level_mark='level1',
          card_mark='onecard',
          essential_mark='unessential')
def test_tensor_sigmoid_dynamic():
    """
    Feature: Test sigmoid op.
    Description: Test sigmoid dynamic shape.
    Expectation: the result match with expected result.
    """
    ms_data1 = ms.Tensor(generate_random_input((4, 3, 6), np.float32))
    ms_data2 = ms.Tensor(generate_random_input((5, 2, 7, 3), np.float32))
    TEST_OP(sigmoid_forward_func, [[ms_data1], [ms_data2]],
            disable_mode=['GRAPH_MODE_GE'])

File: tests/ut/cpp/backend/ms_backend/graph_fusion/expander/test_sigmoid.cc
/**
 * Copyright 2024-2025 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <map>
#include <string>
#include <iostream>
#include "backend/ms_backend/graph_fusion/common/graph_kernel_common_test_suite.h"
#include "utils/anf_utils.h"
#include "abstract/abstract_value.h"
#include "common/graph_optimizer_test_framework.h"
#include "include/runtime/hardware_abstract/kernel_base/graph_fusion/graph_kernel_flags.h"
#include "backend/ms_backend/graph_fusion/adapter/graph_kernel_expander_cloud.h"
#include "include/backend/anf_runtime_algorithm.h"
#include "pre_activate/common/pattern_to_pattern_pass_utils.h"
#include "backend/ms_backend/graph_fusion/expander/base.h"
#include "mindspore/ops/op_def/auto_generate/gen_ops_primitive_c.h"
#include "ir/graph_utils.h"

namespace mindspore::graphkernel::test {
namespace {
struct SigmoidParams {
  bool can_expand;
  ShapeVector input_shape;
  ShapeVector expect_shape;
  TypePtr type;
};

struct GradParams {
  bool can_expand;
  TypePtr type;
  ShapeVector shape;
};
}  // namespace

/// Feature: Test graph kernel Sigmoid expander
/// Description: Sigmoid will expanded
/// Expectation: After expand, the output shape and data type of sub graph should match expect
class TestSigmoidExpander : public TestGraphKernelExpander, public testing::WithParamInterface<SigmoidParams> {
  void SetUp() override {
    SetDeviceTarget(kAscendDevice);
    std::map<std::string, std::string> jit_config;
    jit_config["graph_kernel_flags"] = "--kernel_generator=DVM";
    graphkernel::GraphKernelFlags::SaveJitConfig(jit_config);
  }
};

TEST_P(TestSigmoidExpander, Sigmoid) {
  const auto &param = GetParam();
  ConstructGraph c;
  auto shape = c.NewTensorInput("input_shape", param.type, param.input_shape);
  auto op = c.NewCNodeWithBuildInfo("Sigmoid", {shape}, {});
  c.SetOutput(op);
  RunPass(c.GetGraph(), {std::make_shared<graphkernel::GraphKernelExpanderCloud>()});
  auto nodes = TopoSort(c.GetGraph()->get_return());
  for (const auto &node : nodes) {
    if (node != nullptr && AnfUtils::IsGraphKernel(node)) {
      auto fg = GetCNodeFuncGraph(node);
      auto output_node = fg->output();
      if (IsPrimitiveCNode(output_node, prim::kPrimCast)) {
        auto cnode = output_node->cast<CNodePtr>();
        MS_EXCEPTION_IF_NULL(cnode);
        output_node = cnode->input(1);
      }
      CompareShapeAndType(output_node, 0, param.expect_shape, kFloat32->type_id());
    }
  }
  size_t gk_size = param.can_expand ? 1 : 0;
  ASSERT_EQ(GetAllGKNodes(c.GetGraph()).size(), gk_size);
}

/// Feature: Test graph kernel SigmoidGrad expander
/// Description: test op with different inputs
/// Expectation: Can be expanded only when its input data types are supported.
class TestSigmoidGradExpander : public TestGraphKernelExpander, public testing::WithParamInterface<GradParams> {
  void SetUp() override { SetDeviceTarget(kAscendDevice); }
};

TEST_P(TestSigmoidGradExpander, sigmoid_grad) {
  const auto &param = GetParam();
  ConstructGraph c;
  auto y = c.NewTensorInput("y", param.type, param.shape);
  auto dy = c.NewTensorInput("dy", param.type, param.shape);
  auto op = c.NewCNodeWithBuildInfo("SigmoidGrad", {y, dy});
  c.SetOutput(op);
  RunPass(c.GetGraph(), {std::make_shared<graphkernel::GraphKernelExpanderCloud>()});
  size_t gk_size = param.can_expand ? 1 : 0;
  ASSERT_EQ(GetAllGKNodes(c.GetGraph()).size(), gk_size);
}

INSTANTIATE_TEST_CASE_P(TestOpSigmoid, TestSigmoidExpander,
                        testing::Values(SigmoidParams{true, {16, 16}, {16, 16}, kFloat16},
                                        SigmoidParams{true, {16, 16}, {16, 16}, kFloat32},
                                        SigmoidParams{true, {16, 16}, {16, 16}, kBFloat16},
                                        SigmoidParams{false, {16, 16}, {16, 16}, kInt64}));

INSTANTIATE_TEST_CASE_P(TestOpSigmoidGrad, TestSigmoidGradExpander,
                        testing::Values(GradParams{true, kFloat16, {16, 16}}, GradParams{true, kFloat32, {16, 16}},
                                        GradParams{true, kBFloat16, {16, 16}}, GradParams{true, kBFloat16, {-1, -1}},
                                        GradParams{false, kInt64, {16, 16}}));
}  // namespace mindspore::graphkernel::test